File: app/src/main/java/com/klinker/android/twitter_l/data/roomdb/SavedTweet.java
package com.klinker.android.twitter_l.data.roomdb;


import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.Index;
import androidx.room.PrimaryKey;
import twitter4j.Status;


@Entity(tableName = "saved_tweets",
        indices = { @Index(value = {"tweet_id", "account"}, unique = true)},
        foreignKeys = { @ForeignKey(entity = Tweet.class, parentColumns = "id", childColumns = "tweet_id", onDelete = ForeignKey.CASCADE)})
public class SavedTweet {

    @PrimaryKey(autoGenerate = true)
    public Long id;

    @ColumnInfo
    public int account;

    @ColumnInfo(name="tweet_id")
    public long tweetId;

    public SavedTweet(long id, int account, long tweetId) {
        this.id = id;
        this.account = account;
        this.tweetId = tweetId;
    }

    public SavedTweet(Status status, int account) {
        this.account = account;
        this.tweetId = status.getId();
    }

}

File: app/src/main/java/com/klinker/android/twitter_l/data/roomdb/SavedTweet.kt
package com.klinker.android.twitter_l.data.roomdb

import androidx.room.*
import twitter4j.Status


@Entity(tableName = "saved_tweets")
data class SavedTweet @JvmOverloads constructor(
        @Embedded val tweet: Tweet,
        val account: Int,
        @PrimaryKey(autoGenerate = true) val id: Long = 0
) : BaseDao.TalonEntity<SavedTweet> {
    constructor(status: Status, account: Int) : this(Tweet(status), account)

    override fun copyWithId(id: Long): SavedTweet {
        return this.copy(id = id)
    }
}

@Dao
abstract class SavedTweetDao : BaseDao<SavedTweet>() {

    @Query("SELECT * FROM saved_tweets WHERE account = :account ORDER BY tweet_id ASC LIMIT :pageSize OFFSET ((:page - 1) * :pageSize)")
    abstract fun getSavedTweets(account: Int, page: Int = 1, pageSize: Int = 200) : List<SavedTweet>

    @Ignore
    fun create(status: Status, account: Int) : SavedTweet {
        return this.insert(SavedTweet(status, account))
    }

}